File: src/libadapters.c
// Copyright 2024 Roy T. Hashimoto. All Rights Reserved.
#include <stdio.h>
#include <string.h>
#include <strings.h>
#include <emscripten.h>
#include <sqlite3.h>

// Declarations for synchronous and asynchronous JavaScript relay methods.
// The function name contains the C signature of the JavaScript function.
// The first two arguments of each relay method is the target (e.g. VFS)
// and method name (e.g. xOpen) to call. The remaining arguments are the
// parameters to the method.
//
// Relaying is necessary because Emscripten only allows calling a statically
// defined JavaScript function via a C function pointer.
#define P const void*
#define I int
#define J int64_t
#define DECLARE(TYPE, NAME, ...) \
  extern TYPE NAME(__VA_ARGS__); \
  extern TYPE NAME##_async(__VA_ARGS__);

DECLARE(I, ippp, P, P, P);
DECLARE(void, vppp, P, P, P);
DECLARE(I, ipppj, P, P, P, J);
DECLARE(I, ipppi, P, P, P, I);
DECLARE(I, ipppp, P, P, P, P);
DECLARE(I, ipppip, P, P, P, I, P);
DECLARE(I, ippppi, P, P, P, P, I);
DECLARE(I, ipppiii, P, P, P, I, I, I);
DECLARE(I, ippppij, P, P, P, P, I, J);
DECLARE(I, ippppip, P, P, P, P, I, P);
DECLARE(I, ipppppip, P, P, P, P, P, I, P);
DECLARE(I, ipppiiip, P, P, P, I, I, I, P);
#undef DECLARE
#undef P
#undef I
#undef J

// This list of methods must match exactly with libadapters.js.
enum {
  xOpen,
  xDelete,
  xAccess,
  xFullPathname,
  xRandomness,
  xSleep,
  xCurrentTime,
  xGetLastError,
  xCurrentTimeInt64,

  xClose,
  xRead,
  xWrite,
  xTruncate,
  xSync,
  xFileSize,
  xLock,
  xUnlock,
  xCheckReservedLock,
  xFileControl,
  xSectorSize,
  xDeviceCharacteristics,
  xShmMap,
  xShmLock,
  xShmBarrier,
  xShmUnmap
};

// Attach extra information to the VFS and file objects.
typedef struct VFS {
  sqlite3_vfs base;
  int methodMask; // Bitmask of methods defined in JavaScript.
  int asyncMask;  // Bitmask of methods that are asynchronous.
} VFS;

typedef struct VFSFile {
  sqlite3_file base;
  VFS* pVfs; // Pointer back to the VFS.
} VFSFile;

#define VFS_JS(SIGNATURE, KEY, METHOD, ...) \
  (((VFS*)KEY)->asyncMask & (1 << METHOD) ? \
    SIGNATURE##_async(KEY, #METHOD, __VA_ARGS__) : \
    SIGNATURE(KEY, #METHOD, __VA_ARGS__))

static int adapter_xClose(sqlite3_file* pFile) {
  return VFS_JS(ippp, ((VFSFile*)pFile)->pVfs, xClose, pFile);
}

static int adapter_xRead(sqlite3_file* pFile, void* pData, int iAmt, sqlite3_int64 iOffset) {
  return VFS_JS(ippppij, ((VFSFile*)pFile)->pVfs, xRead, pFile, pData, iAmt, iOffset);
}

static int adapter_xWrite(sqlite3_file* pFile, const void* pData, int iAmt, sqlite3_int64 iOffset) {
  return VFS_JS(ippppij, ((VFSFile*)pFile)->pVfs, xWrite, pFile, pData, iAmt, iOffset);
}

static int adapter_xTruncate(sqlite3_file* pFile, sqlite3_int64 size) {
  return VFS_JS(ipppj, ((VFSFile*)pFile)->pVfs, xTruncate, pFile, size);
}

static int adapter_xSync(sqlite3_file* pFile, int flags) {
  return VFS_JS(ipppi, ((VFSFile*)pFile)->pVfs, xSync, pFile, flags);
}

static int adapter_xFileSize(sqlite3_file* pFile, sqlite3_int64* pSize) {
  return VFS_JS(ipppp, ((VFSFile*)pFile)->pVfs, xFileSize, pFile, pSize);
}

static int adapter_xLock(sqlite3_file* pFile, int lockType) {
  return VFS_JS(ipppi, ((VFSFile*)pFile)->pVfs, xLock, pFile, lockType);
}

static int adapter_xUnlock(sqlite3_file* pFile, int lockType) {
  return VFS_JS(ipppi, ((VFSFile*)pFile)->pVfs, xUnlock, pFile, lockType);
}

static int adapter_xCheckReservedLock(sqlite3_file* pFile, int* pResOut) {
  return VFS_JS(ipppp, ((VFSFile*)pFile)->pVfs, xCheckReservedLock, pFile, pResOut);
}

static int adapter_xFileControl(sqlite3_file* pFile, int flags, void* pOut) {
  return VFS_JS(ipppip, ((VFSFile*)pFile)->pVfs, xFileControl, pFile, flags, pOut);
}

static int adapter_xSectorSize(sqlite3_file* pFile) {
  return VFS_JS(ippp, ((VFSFile*)pFile)->pVfs, xSectorSize, pFile);
}

static int adapter_xDeviceCharacteristics(sqlite3_file* pFile) {
  return VFS_JS(ippp, ((VFSFile*)pFile)->pVfs, xDeviceCharacteristics, pFile);
}

static int adapter_xShmMap(sqlite3_file* pFile, int iPg, int pgsz, int unused, void volatile** p) {
  return VFS_JS(ipppiiip, ((VFSFile*)pFile)->pVfs, xShmMap, pFile, iPg, pgsz, unused, p);
}

static int adapter_xShmLock(sqlite3_file* pFile, int offset, int n, int flags) {
  return VFS_JS(ipppiii, ((VFSFile*)pFile)->pVfs, xShmLock, pFile, offset, n, flags);
}

static void adapter_xShmBarrier(sqlite3_file* pFile) {
  VFS_JS(vppp, ((VFSFile*)pFile)->pVfs, xShmBarrier, pFile);
}

static int adapter_xShmUnmap(sqlite3_file* pFile, int deleteFlag) {
  return VFS_JS(ipppi, ((VFSFile*)pFile)->pVfs, xShmUnmap, pFile, deleteFlag);
}


static int adapter_xOpen(sqlite3_vfs* pVfs, const char* zName, sqlite3_file* pFile, int flags, int* pOutFlags) {
  const int result = VFS_JS(ipppppip, pVfs, xOpen, pVfs, (void*)zName, pFile, flags, pOutFlags);

  VFS* pVfsExt = (VFS*)pVfs;
  sqlite3_io_methods* pMethods = (sqlite3_io_methods*)sqlite3_malloc(sizeof(sqlite3_io_methods));
  pMethods->iVersion = 2;
#define METHOD(NAME) pMethods->NAME = (pVfsExt->methodMask & (1 << NAME)) ? adapter_##NAME : NULL
  METHOD(xClose);
  METHOD(xRead);
  METHOD(xWrite);
  METHOD(xTruncate);
  METHOD(xSync);
  METHOD(xFileSize);
  METHOD(xLock);
  METHOD(xUnlock);
  METHOD(xCheckReservedLock);
  METHOD(xFileControl);
  METHOD(xSectorSize);
  METHOD(xDeviceCharacteristics);
  METHOD(xShmMap);
  METHOD(xShmLock);
  METHOD(xShmBarrier);
  METHOD(xShmUnmap);
#undef METHOD
  pFile->pMethods = pMethods;
  ((VFSFile*)pFile)->pVfs = pVfsExt;
  return result;
}

static int adapter_xDelete(sqlite3_vfs* pVfs, const char* zName, int syncDir) {
  return VFS_JS(ippppi, pVfs, xDelete, pVfs, zName, syncDir);
}

static int adapter_xAccess(sqlite3_vfs* pVfs, const char* zName, int flags, int* pResOut) {
  return VFS_JS(ippppip, pVfs, xAccess, pVfs, zName, flags, pResOut);
}

static int adapter_xFullPathname(sqlite3_vfs* pVfs, const char* zName, int nOut, char* zOut) {
  return VFS_JS(ippppip, pVfs, xFullPathname, pVfs, zName, nOut, zOut);
}

static int adapter_xRandomness(sqlite3_vfs* pVfs, int nBuf, char* zBuf) {
  return VFS_JS(ipppip, pVfs, xRandomness, pVfs, nBuf, zBuf);
}

static int adapter_xSleep(sqlite3_vfs* pVfs, int microseconds) {
  return VFS_JS(ipppi, pVfs, xSleep, pVfs, microseconds);
}

static int adapter_xCurrentTime(sqlite3_vfs* pVfs, double* pJulianDay) {
  return VFS_JS(ipppp, pVfs, xCurrentTime, pVfs, pJulianDay);
}

static int adapter_xGetLastError(sqlite3_vfs* pVfs, int nBuf, char* zBuf) {
  return VFS_JS(ipppip, pVfs, xGetLastError, pVfs, nBuf, zBuf);
}

static int adapter_xCurrentTimeInt64(sqlite3_vfs* pVfs, sqlite3_int64* pTime) {
  return VFS_JS(ipppp, pVfs, xCurrentTimeInt64, pVfs, pTime);
}

int EMSCRIPTEN_KEEPALIVE adapter_vfs_register(
  const char* zName,
  int mxPathName,
  int methodMask,
  int asyncMask,
  int makeDefault,
  void** ppVfs) {
  // Get the current default VFS to use if methods are not defined.
  const sqlite3_vfs* backupVfs = sqlite3_vfs_find(NULL);

  // Allocate and populate the new VFS.
  VFS* vfs = (VFS*)sqlite3_malloc(sizeof(VFS));
  if (!vfs) return SQLITE_NOMEM;
  bzero(vfs, sizeof(VFS));

  vfs->base.iVersion = 2;
  vfs->base.szOsFile = sizeof(VFSFile);
  vfs->base.mxPathname = mxPathName;
  vfs->base.zName = strdup(zName);

  // The VFS methods go to the adapter implementations in this file,
  // or to the default VFS if the JavaScript method is not defined.
#define METHOD(NAME) vfs->base.NAME = \
  (methodMask & (1 << NAME)) ? adapter_##NAME : backupVfs->NAME

  METHOD(xOpen);
  METHOD(xDelete);
  METHOD(xAccess);
  METHOD(xFullPathname);
  METHOD(xRandomness);
  METHOD(xSleep);
  METHOD(xCurrentTime);
  METHOD(xGetLastError);
  METHOD(xCurrentTimeInt64);
#undef METHOD

  vfs->methodMask = methodMask;
  vfs->asyncMask = asyncMask;

  *ppVfs = vfs;
  return sqlite3_vfs_register(&vfs->base, makeDefault);
}

// Some SQLite API functions take a pointer to a function that frees
// memory. Although we could add a C binding to a JavaScript function
// that calls sqlite3_free(), it is more efficient to pass the sqlite3_free
// function pointer directly. This function provides the C pointer to
// JavaScript.
void* EMSCRIPTEN_KEEPALIVE getSqliteFree() {
  return sqlite3_free;
}

int main() {
  sqlite3_initialize();
  return 0;
}

File: src/libadapters.js
// Method names for these signatures must be in src/asyncify_imports.json.
const SIGNATURES = [
  'ippp', // xClose, xSectorSize, xDeviceCharacteristics
  'vppp', // xShmBarrier, xFinal
  'ipppj', // xTruncate
  'ipppi', // xSleep, xSync, xLock, xUnlock, xShmUnmap
  'ipppp', // xFileSize, xCheckReservedLock, xCurrentTime, xCurrentTimeInt64
  'ipppip', // xFileControl, xRandomness, xGetLastError
  'vpppip', // xFunc, xStep
  'ippppi', // xDelete
  'ippppij', // xRead, xWrite
  'ipppiii', // xShmLock
  'ippppip', // xAccess, xFullPathname
  'ipppppip', // xOpen
  'ipppiiip', // xShmMap
];

// @ts-ignore
// This object will define the methods callable from WebAssembly.
// See https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#implement-a-c-api-in-javascript
//
// At this writing, asynchronous JavaScript functions to be called from
// WebAssembly must be statically defined, i.e. they cannot be registered
// at runtime. The workaround here is to define synchronous and asynchronous
// relaying functions for each needed call signature.
//
// On the C side, calls are made to the relaying function with two prepended
// arguments (key, methodName). The relaying function then looks up and
// calls the appropriate receiver and method.
const adapters = {
  $adapters_support: function() {
    const AsyncFunction = Object.getPrototypeOf(async function(){}).constructor;

    // @ts-ignore
    // Expose handleAsync to library and application code.
    const handleAsync = typeof Asyncify === 'object' ?
      Asyncify.handleAsync.bind(Asyncify) :
      null;
    Module['handleAsync'] = handleAsync;

    // This map contains the objects to which calls will be relayed, e.g.
    // a VFS. The key is typically the corresponding WebAssembly pointer.
    const targets = new Map();

    // @ts-ignore
    // Overwrite this function with the relay service function.
    adapters_support = function(isAsync, key, ...args) {
      // If the receiver found with the key is a function, just call it.
      // Otherwise, the next argument is the name of the method to be called.
      const receiver = targets.get(key);
      let methodName = null;
      const f = typeof receiver === 'function' ?
        receiver :
        receiver[methodName = UTF8ToString(args.shift())];
      
      if (isAsync) {
        // Call async function via handleAsync. This works for both
        // Asyncify and JSPI builds.
        if (handleAsync) {
          return handleAsync(() => f.apply(receiver, args));
        }
        throw new Error('Synchronous WebAssembly cannot call async function');
      }

      // The function should not be async so call it directly.
      const result = f.apply(receiver, args);
      if (typeof result?.then == 'function') {
        console.error('unexpected Promise', f);
        throw new Error(`${methodName} unexpectedly returned a Promise`);
      }
      return result;
    };

    // This list of methods must match exactly with libadapters.c.
    const VFS_METHODS = [
      'xOpen',
      'xDelete',
      'xAccess',
      'xFullPathname',
      'xRandomness',
      'xSleep',
      'xCurrentTime',
      'xGetLastError',
      'xCurrentTimeInt64',

      'xClose',
      'xRead',
      'xWrite',
      'xTruncate',
      'xSync',
      'xFileSize',
      'xLock',
      'xUnlock',
      'xCheckReservedLock',
      'xFileControl',
      'xSectorSize',
      'xDeviceCharacteristics',
      'xShmMap',
      'xShmLock',
      'xShmBarrier',
      'xShmUnmap'
    ];

    Module['vfs_register'] = function(vfs, makeDefault) {
      // Determine which methods exist and which are asynchronous.
      let methodMask = 0;
      let asyncMask = 0;
      const AsyncFunction = Object.getPrototypeOf(async function(){}).constructor;
      VFS_METHODS.forEach((method, i) => {
        if (vfs[method]) {
          methodMask |= 1 << i;
          if (vfs['hasAsyncMethod'](method)) {
            asyncMask |= 1 << i;
          }
        }
      });

      // Allocate space for adapter_vfs_register to write the sqlite3_vfs
      // pointer. This pointer will be used to look up the JavaScript VFS
      // object.
      const vfsPointer = Module['_malloc'](4);
      try {
        const result = ccall(
          'adapter_vfs_register',
          'number',
          ['string', 'number', 'number', 'number', 'number', 'number'],
          [vfs.name, vfs.mxPathname, methodMask, asyncMask, makeDefault ? 1 : 0, vfsPointer]);
        if (!result) {
          const key = getValue(vfsPointer, '*');
          targets.set(key, vfs);
        }
        return result;
      } finally {
        Module['_free'](vfsPointer);
      }
    };

    const FUNC_METHODS = [
      'xFunc',
      'xStep',
      'xFinal'
    ];

    const mapFunctionNameToKey = new Map();

    Module['create_function'] = function(db, zFunctionName, nArg, eTextRep, pApp, xFunc, xStep, xFinal) {
      // Allocate some memory to store the async flags. In addition, this
      // pointer is passed to SQLite as the application data (the user's
      // application data is ignored), and is used to look up the JavaScript
      // target object.
      const pAsyncFlags = Module['_sqlite3_malloc'](4);
      const target = { xFunc, xStep, xFinal };
      setValue(pAsyncFlags, FUNC_METHODS.reduce((mask, method, i) => {
        if (target[method] instanceof AsyncFunction) {
          return mask | 1 << i;
        }
        return mask;
      }, 0), 'i32');

      const result = ccall(
        'adapter_create_function',
        'number',
        ['number', 'string', 'number', 'number', 'number', 'number', 'number', 'number'],
        [
          db,
          zFunctionName,
          nArg,
          eTextRep,
          pAsyncFlags,
          xFunc ? 1 : 0,
          xStep ? 1 : 0,
          xFinal? 1 : 0
        ]);
      if (!result) {
        if (mapFunctionNameToKey.has(zFunctionName)) {
          // Reclaim the old resources used with this name.
          const oldKey = mapFunctionNameToKey.get(zFunctionName);
          targets.delete(oldKey);
        }
        mapFunctionNameToKey.set(zFunctionName, pAsyncFlags);
        targets.set(pAsyncFlags, { xFunc, xStep, xFinal });
      }
      return result;
    };
  },
  $adapters_support__deps: ['$UTF8ToString'],
  $adapters_support__postset: 'adapters_support();',
};

function injectMethod(signature, isAsync) {
  const method = `${signature}${isAsync ? '_async' : ''}`;
  adapters[`${method}`] = isAsync ?
    // @ts-ignore
    function(...args) { return adapters_support(true, ...args) } :
    // @ts-ignore
    function(...args) { return adapters_support(false, ...args) };
  adapters[`${method}__deps`] = ['$adapters_support'];
  adapters[`${method}__async`] = isAsync;

  // Emscripten "legalizes" 64-bit integer arguments by passing them as
  // two 32-bit signed integers.
  adapters[`${method}__sig`] = `${signature[0]}${signature.substring(1).replaceAll('j', 'ii')}`;
}

// For each function signature, inject a synchronous and asynchronous
// relaying method definition.
for (const signature of SIGNATURES) {
  injectMethod(signature, false);
  injectMethod(signature, true);
}

// @ts-ignore
addToLibrary(adapters);